### PR TITLE
Split out signalflow into separate module

### PR DESCRIPTION
**Pull request by breedx-splk**
This modularizes the code quite a bit. There are some tangled dependencies, so 2 additional modules also needed to be factored out: `signalfx-connection` and `signalfx-endpoint`.

The dependency graph is basically:

```mermaid
graph TD;
    signalfx-connection-->signalfx-endpoint;
    signalfx-signalflow-->signalfx-endpoint;
    signalfx-signalflow-->signalfx-connection;
    signalfx-java-->signalfx-connection;
    signalfx-java-->signalfx-endpoint;
    signalfx-java-->signalfx-signalflow;
```

There was one code change needed to eliminate cycles there. The `SignalFxMetricsException` which was an unnecessary (named/tag) subclass of `RuntimeException` was replaced with just `RuntimeException`.

While I do think this split helps to better organize and show where dependencies are and to eventually pull out signalflow, it also highlights that signalflow still has some dependencies on this common stuff.

### CODE REVIEW

File: signalfx-connection/pom.xml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    </dependency>
+    <!-- test -->
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.httpcomponents</groupId>
+      <artifactId>httpclient</artifactId>
+    </dependency>
```
Comment (mateuszrzeszutek) [inline signalfx-connection/pom.xml:66]: That's defined as a compile dependency in the parent pom

File: signalfx-connection/pom.xml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  <parent>
+    <groupId>com.signalfx.public</groupId>
+    <artifactId>clients-parent</artifactId>
+    <version>1.0.30</version>
+  </parent>
+
+  <properties>
+    <!-- This prevents shading from happening in this module -->
+    <skipShaded>true</skipShaded>
+    <maven.deploy.skip>false</maven.deploy.skip>
```
Comment (mateuszrzeszutek) [inline signalfx-connection/pom.xml:16]: Just wondering, what's this for?

File: pom.xml [hunk 3]
```diff
@@ -176,11 +177,11 @@
...
-      <dependency>
-        <groupId>org.eclipse.jetty.websocket</groupId>
-        <artifactId>websocket-client</artifactId>
-        <version>${jetty.version}</version>
-      </dependency>
+<!--      <dependency>-->
+<!--        <groupId>org.eclipse.jetty.websocket</groupId>-->
+<!--        <artifactId>websocket-client</artifactId>-->
+<!--        <version>${jetty.version}</version>-->
+<!--      </dependency>-->
```
Comment (mateuszrzeszutek) [inline pom.xml:184]: To be removed?
```suggestion
```

File: signalfx-signalflow/pom.xml [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    </dependency>
+    <dependency>
+      <groupId>org.eclipse.jetty.websocket</groupId>
+      <artifactId>websocket-client</artifactId>
+      <version>${jetty.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.eclipse.jetty</groupId>
+      <artifactId>jetty-server</artifactId>
+    </dependency>
```
Comment (mateuszrzeszutek) [inline signalfx-signalflow/pom.xml:102]: That's a test dependency actually

File: signalfx-java/pom.xml [hunk 5]
```diff
@@ -43,15 +43,37 @@
...
+      <groupId>com.signalfx.public</groupId>
+      <artifactId>signalfx-connection</artifactId>
+      <version>${version}</version>
+    </dependency>
     <dependency>
-      <groupId>org.eclipse.jetty.websocket</groupId>
-      <artifactId>websocket-client</artifactId>
+      <groupId>com.signalfx.public</groupId>
+      <artifactId>signalfx-endpoint</artifactId>
+      <version>${version}</version>
```
Comment (mateuszrzeszutek) [inline signalfx-java/pom.xml:55]: These two also appear in the `<!-- compile -->` section; in an identical form.

**Review by mateuszrzeszutek — COMMENTED** [review]
Awesome!

File: signalfx-connection/pom.xml (hunk 1 above)
Comment (breedx-splk) [reply signalfx-connection/pom.xml:66]: I think the parent just sets up the versions...so this is definitely required, I confirmed that removing it breaks the build.

File: signalfx-connection/pom.xml (hunk 2 above)
Comment (breedx-splk) [reply signalfx-connection/pom.xml:16]: It's wonky.  So without this (and the new setting in the parent pom), every child module/jar gets its dependencies shaded, which can result in public/external APIs having breaking changes to their interface. So a method that normally takes something like `org.apache.FooImpl` gets changed to `com.signalfx.shaded.org.apache.FooImpl` and this breaks all the callers. I couldn't find another way around that, unfortunately.

File: signalfx-java/pom.xml (hunk 5 above)
Comment (breedx-splk) [reply signalfx-java/pom.xml:55]: Thanks!

**Comment by breedx-splk** [comment]
ready for another look @mateuszrzeszutek thanks!

File: signalfx-connection/pom.xml (hunk 2 above)
Comment (mateuszrzeszutek) [reply signalfx-connection/pom.xml:16]: OOF
maven magic 🪄

File: signalfx-connection/pom.xml (hunk 1 above)
Comment (mateuszrzeszutek) [reply signalfx-connection/pom.xml:66]: > I think the parent just sets up the versions...so this is definitely required, I confirmed that removing it breaks the build.

I'm not saying that it's not required; I'm saying that if it's a test dep it needs `<scope>test</scope>`

Comment (breedx-splk) [reply signalfx-connection/pom.xml:66]: Yeah, that doesn't work because it's used by the `RetryStrategy` and `RetryHandler` which are production classes.

**Review by mateuszrzeszutek — APPROVED** [review]
🎉

**Pull request merged**